File: library/src/main/java/com/androidwind/androidquick/module/retrofit/RetrofitManager.java
package com.androidwind.androidquick.module.retrofit;

import com.androidwind.androidquick.BuildConfig;
import com.androidwind.androidquick.util.LogUtil;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * @author ddnosh
 * @website http://blog.csdn.net/ddnosh
 */
public class RetrofitManager {

    private static final String TAG = "RetrofitManager";

    private Retrofit retrofit;
    private Retrofit.Builder retrofitBuilder;
    private OkHttpClient okHttpClient;
    private OkHttpClient.Builder okHttpClientBuilder;

    private static String BASE_URL = "https://api.github.com";

    private static RetrofitManager mInstance;

    public static RetrofitManager getInstance() {
        if (mInstance == null) {
            synchronized (RetrofitManager.class) {
                if (mInstance == null) {
                    mInstance = new RetrofitManager(BASE_URL);
                }
            }
        }
        return mInstance;
    }

    public RetrofitManager(String baseUrl) {
        BASE_URL = baseUrl;
        initOkHttp();
        LogUtil.i(TAG, "RetrofitManager ->" + BASE_URL);
    }

    public static void initBaseUrl(String url) {
        BASE_URL = url;
        LogUtil.i(TAG, "initBaseUrl ->" + BASE_URL);
    }

    private void initOkHttp() {
        okHttpClientBuilder = new OkHttpClient.Builder();
        if (BuildConfig.DEBUG) {
            // Log信息拦截器
            HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
            loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);//这里可以选择拦截级别
            //设置 Debug Log 模式
            okHttpClientBuilder.addInterceptor(loggingInterceptor);
            //配置SSL证书检测
            okHttpClientBuilder.sslSocketFactory(SSLSocketClient.getNoSSLSocketFactory());
            okHttpClientBuilder.hostnameVerifier(SSLSocketClient.getHostnameVerifier());
        }
        //错误重连
        okHttpClientBuilder.retryOnConnectionFailure(true);
        okHttpClient = okHttpClientBuilder.build();
        LogUtil.i(TAG, "initOkHttp:getNoSSLSocketFactory");
    }

    public <T> T createApi(Class<T> clazz) {
        if (retrofit == null) {
            synchronized (RetrofitManager.class) {
                if (retrofit == null) {
                    //Retrofit
                    retrofitBuilder = new Retrofit.Builder();
                    retrofitBuilder.baseUrl(BASE_URL)
                            .client(okHttpClient)
                            .addConverterFactory(GsonConverterFactory.create())//定义转化器,用Gson将服务器返回的Json格式解析成实体
                            .addCallAdapterFactory(RxJava2CallAdapterFactory.create());//关联RxJava
                    retrofit = retrofitBuilder.build();
                }
            }
        }
        LogUtil.i(TAG, "initRetrofit");
        return retrofit.create(clazz);
    }

    public Retrofit getRetrofit() {
        return retrofit;
    }

    public void setRetrofit(Retrofit retrofit) {
        this.retrofit = retrofit;
    }

    public Retrofit.Builder getRetrofitBuilder() {
        return retrofitBuilder;
    }

    public void setRetrofitBuilder(Retrofit.Builder retrofitBuilder) {
        this.retrofitBuilder = retrofitBuilder;
    }

    public OkHttpClient getOkHttpClient() {
        return okHttpClient;
    }

    public void setOkHttpClient(OkHttpClient okHttpClient) {
        this.okHttpClient = okHttpClient;
    }

    public OkHttpClient.Builder getOkHttpClientBuilder() {
        return okHttpClientBuilder;
    }

    public void setOkHttpClientBuilder(OkHttpClient.Builder okHttpClientBuilder) {
        this.okHttpClientBuilder = okHttpClientBuilder;
    }
}

File: app/src/main/java/com/androidwind/androidquick/sample/http/RetrofitManager.kt
package com.androidwind.androidquick.sample.http

import com.androidwind.androidquick.sample.AppConfig
import com.androidwind.androidquick.sample.MyApplication
import com.androidwind.androidquick.util.FileUtil
import okhttp3.Cache
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.converter.simplexml.SimpleXmlConverterFactory
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * @author ddnosh
 * @website http://blog.csdn.net/ddnosh
 */
enum class RetrofitManager {
    INSTANCE;

    private val TAG = "RetrofitManager"
    private val retrofitMap = HashMap<String, Retrofit?>()
    private fun createRetrofit(baseUrl: String, isJson: Boolean) {
        val timeOut = AppConfig.HTTP_TIME_OUT
        val cache = Cache(MyApplication.instance?.let { FileUtil.getHttpImageCacheDir(it) }!!,
            AppConfig.HTTP_MAX_CACHE_SIZE.toLong())
        // Log信息拦截器
        val loggingInterceptor = HttpLoggingInterceptor()
        loggingInterceptor.level = HttpLoggingInterceptor.Level.BODY //这里可以选择拦截级别
        val okHttpClient = OkHttpClient.Builder()
            .connectTimeout(timeOut.toLong(), TimeUnit.MILLISECONDS)
            .addInterceptor(loggingInterceptor)
            .cache(cache)
            .build()
        val builder = Retrofit.Builder()
            .baseUrl(baseUrl)
            .addConverterFactory(GsonConverterFactory.create()) //定义转化器,用Gson将服务器返回的Json格式解析成实体
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create()) //关联RxJava
            .client(okHttpClient)
        if (isJson) {
            builder.addConverterFactory(GsonConverterFactory.create())
        } else {
            builder.addConverterFactory(SimpleXmlConverterFactory.createNonStrict())
        }
        retrofitMap["$baseUrl-$isJson"] = builder.build()
    }

    fun getRetrofit(baseUrl: String, isJson: Boolean): Retrofit? {
        val key = "$baseUrl-$isJson"
        if (!retrofitMap.containsKey(key)) {
            createRetrofit(baseUrl, isJson)
        }
        return retrofitMap[key]
    }

    fun getRetrofit(baseUrl: String): Retrofit? {
        return getRetrofit(baseUrl, true)
    }
}